src/planner/planners/MatchVariableLengthPatternIndexScanPlanner.cpp

```diff
@@ -0,0 +1,509 @@
+    auto scan = IndexScan::make(matchCtx_->qctx,
+                                nullptr,
+                                matchCtx_->space.id,
+                                std::move(contexts),
+                                std::move(columns),
+                                false,
+                                matchCtx_->scanInfo.schemaId);
+    plan->tail = scan;
+    plan->root = scan;
+
+    // initialize start expression in project node
+    initialExpr_ = ExpressionUtils::newVarPropExpr(kVid);
+
+    return Status::OK();
+}
+
+PlanNode *MatchVariableLengthPatternIndexScanPlanner::joinDataSet(const PlanNode *right,
+                                                                  const PlanNode *left) {
+    auto &leftKey = left->colNamesRef().back();
+    auto &rightKey = right->colNamesRef().front();
+    auto buildExpr = getLastEdgeDstExprInLastPath(leftKey);
+    auto probeExpr = getFirstVertexVidInFistPath(rightKey);
+    auto join = DataJoin::make(matchCtx_->qctx,
+                               const_cast<PlanNode *>(right),
+                               {left->outputVar(), 0},
+                               {right->outputVar(), 0},
+                               {buildExpr},
+                               {probeExpr});
+    std::vector<std::string> colNames = left->colNames();
+    const auto &rightColNames = right->colNamesRef();
+    colNames.insert(colNames.end(), rightColNames.begin(), rightColNames.end());
+    join->setColNames(std::move(colNames));
+    return join;
+}
+
+Status MatchVariableLengthPatternIndexScanPlanner::appendFetchVertexPlan(
+    const Expression *nodeFilter,
+    SubPlan *plan) {
+    auto qctx = matchCtx_->qctx;
+
+    extractAndDedupVidColumn(plan);
+    auto srcExpr = ExpressionUtils::inputPropExpr(kVid);
+    auto gv = GetVertices::make(qctx, plan->root, matchCtx_->space.id, srcExpr.release(), {}, {});
+
+    PlanNode *root = gv;
+    if (nodeFilter != nullptr) {
+        auto filter = nodeFilter->clone().release();
+        RewriteMatchLabelVisitor visitor([](const Expression *expr) {
+            DCHECK_EQ(expr->kind(), Expression::Kind::kLabelAttribute);
+            auto la = static_cast<const LabelAttributeExpression *>(expr);
+            auto attr = new LabelExpression(*la->right()->name());
+            return new AttributeExpression(new VertexExpression(), attr);
+        });
+        filter->accept(&visitor);
+        root = Filter::make(matchCtx_->qctx, root, filter);
+    }
+
+    // normalize all columns to one
+    auto columns = saveObject(new YieldColumns);
+    auto pathExpr = std::make_unique<PathBuildExpression>();
+    pathExpr->add(std::make_unique<VertexExpression>());
+    columns->addColumn(new YieldColumn(pathExpr.release()));
+    plan->root = Project::make(qctx, root, columns);
+    plan->root->setColNames({kPath});
+    return Status::OK();
+}
+
+Status MatchVariableLengthPatternIndexScanPlanner::filterDatasetByPathLength(const NodeInfo &node,
```

Comment:
Don't need to filter if don't union the data less than min hop.

Reply:
Right, But I don't want to optimize this case here. If so, our plan generated code will be filled by many branches. I prefer to add some optimization rules to handle it.

Reply:
Make sense, but this pattern seems not easy to match in optimizer.